### PR TITLE
Process RepoCred operation

### PR DESCRIPTION
**Pull request by drpaneas**
#### Description:

- This PR handles the logic from the side of cluster-agent. This is getting triggered when an operation CR is present.

#### Link to JIRA Story (if applicable):

(https://issues.redhat.com/browse/GITOPSRVCE-108)

#### For the reviewer:

Here 5 scenarios for the unit-tests.
You can run them: `go test -v ./controllers/managed-gitops/eventloop/... -args --ginkgo.focus "Credentials"` altogether or each scenario separately.

**Scenario 1**
**When** Operation DB row and its corresponding Operation CR, point to an existing RepositoryCredentials DB row, It **then** should create an ArgoCD Secret for Private Repository based on the RepositoryCredentials DB row data

Logs:

```
INFO	Retrieved RepositoryCredentials DB row	{"operationRow": "test-operation", "repositoryCredentialsRow": "test-my-repo-creds"}
INFO	Argo CD Private Repository secret doesn't exist	{"operationRow": "test-operation", "repositoryCredentialsRow": "test-my-repo-creds", "secret": "test-fake-secret-obj", "namespace": "gitops-service-argocd"}
INFO	Argo CD Private Repository secret has been successfully created	{"operationRow": "test-operation", "repositoryCredentialsRow": "test-my-repo-creds", "secret": "test-fake-secret-obj", "namespace": "gitops-service-argocd", "URL": "dGVzdC1mYWtlLXByaXZhdGUtdXJs", "username": "dGVzdC1mYWtlLWF1dGgtdXNlcm5hbWU=", "password": "dGVzdC1mYWtlLWF1dGgtcGFzc3dvcmQ=", "SSH Key": ""}
INFO	Secret is missing ArgoCD label! Syncing with database...	{"operationRow": "test-operation", "repositoryCredentialsRow": "test-my-repo-creds", "secret": "test-fake-secret-obj", "namespace": "gitops-service-argocd", "AddLabel": "argocd.argoproj.io/secret-type: repository"}
INFO	Secret is missing DatabaseID label! Syncing with database...	{"operationRow": "test-operation", "repositoryCredentialsRow": "test-my-repo-creds", "secret": "test-fake-secret-obj", "namespace": "gitops-service-argocd", "AddLabel": "databaseID: test-my-repo-creds"}
INFO	Secret is missing ArgoCD annotation! Syncing with database...	{"operationRow": "test-operation", "repositoryCredentialsRow": "test-my-repo-creds", "secret": "test-fake-secret-obj", "namespace": "gitops-service-argocd", "AddAnnotation": "managed-by: argocd.argoproj.io"}
UPDATE "operation" AS "op" SET "instance_id" = 'test-fake-engine-instance', "resource_id" = 'test-my-repo-creds', "operation_owner_user_id" = 'test-repocred-user-id', "resource_type" = 'RepositoryCredentials', "created_on" = '2022-07-11 15:59:48.138219+00:00:00', "last_state_update" = '2022-07-11 15:59:48.17231+00:00:00', "state" = 'Completed', "human_readable_state" = NULL, "seq_id" = 274 WHERE "op"."operation_id" = 'test-operation'
1.657555188174765e+09	INFO	Updated Operation state for 'test-operation', state is: Completed
```

**Scenario 2**
**When** Operation DB row and its corresponding Operation CR, point to an existing RepositoryCredentials DB row, it **then** should update the existing ArgoCD Secret for Private Repository based on the RepositoryCredentials DB row data

Logs: (pay attention at the logs the keys `UpdatedFrom` to `UpdateTo`)

```
INFO	Retrieved RepositoryCredentials DB row	{"operationRow": "test-operation-2", "repositoryCredentialsRow": "test-my-repo-creds-2"}
INFO	A corresponding Argo CD Private Repository secret has already been existing	{"operationRow": "test-operation-2", "repositoryCredentialsRow": "test-my-repo-creds-2", "secret": "test-fake-secret-wrong-values", "namespace": "gitops-service-argocd", "URL": "wrong-url", "username": "wrong-username", "password": "wrong-password", "SSH Key": "wrong-ssh-key"}
INFO	Secret is missing ArgoCD label! Syncing with database...	{"operationRow": "test-operation-2", "repositoryCredentialsRow": "test-my-repo-creds-2", "secret": "test-fake-secret-wrong-values", "namespace": "gitops-service-argocd", "AddLabel": "argocd.argoproj.io/secret-type: repository"}
INFO	Secret is missing DatabaseID label! Syncing with database...	{"operationRow": "test-operation-2", "repositoryCredentialsRow": "test-my-repo-creds-2", "secret": "test-fake-secret-wrong-values", "namespace": "gitops-service-argocd", "AddLabel": "databaseID: test-my-repo-creds-2"}
INFO	Secret is missing ArgoCD annotation! Syncing with database...	{"operationRow": "test-operation-2", "repositoryCredentialsRow": "test-my-repo-creds-2", "secret": "test-fake-secret-wrong-values", "namespace": "gitops-service-argocd", "AddAnnotation": "managed-by: argocd.argoproj.io"}
INFO	Secret has wrong URL! Syncing with database...	{"operationRow": "test-operation-2", "repositoryCredentialsRow": "test-my-repo-creds-2", "secret": "test-fake-secret-wrong-values", "namespace": "gitops-service-argocd", "UpdateFrom": "wrong-url", "UpdateTo": "test-fake-correct-url"}
INFO	Secret has wrong Password! Syncing with database...	{"operationRow": "test-operation-2", "repositoryCredentialsRow": "test-my-repo-creds-2", "secret": "test-fake-secret-wrong-values", "namespace": "gitops-service-argocd", "UpdateFrom": "wrong-password", "UpdateTo": "test-fake-correct-auth-password"}
INFO	Secret has wrong Username! Syncing with database...	{"operationRow": "test-operation-2", "repositoryCredentialsRow": "test-my-repo-creds-2", "secret": "test-fake-secret-wrong-values", "namespace": "gitops-service-argocd", "UpdateFrom": "wrong-username", "UpdateTo": "test-fake-correct-auth-username"}
INFO	Secret has wrong SSH key! Syncing with database...	{"operationRow": "test-operation-2", "repositoryCredentialsRow": "test-my-repo-creds-2", "secret": "test-fake-secret-wrong-values", "namespace": "gitops-service-argocd", "UpdateFrom": "", "UpdateTo": "test-fake-correct-auth-ssh-key"}
UPDATE "operation" AS "op" SET "instance_id" = 'test-fake-engine-instance', "resource_id" = 'test-my-repo-creds-2', "operation_owner_user_id" = 'test-repocred-user-id', "resource_type" = 'RepositoryCredentials', "created_on" = '2022-07-11 16:01:55.173449+00:00:00', "last_state_update" = '2022-07-11 16:01:55.201733+00:00:00', "state" = 'Completed', "human_readable_state" = NULL, "seq_id" = 275 WHERE "op"."operation_id" = 'test-operation-2'
1.657555315203831e+09	INFO	Updated Operation state for 'test-operation-2', state is: Completed
```

**Scenario 3**
**When** Operation DB row and its corresponding Operation CR, point to a missing RepositoryCredentials DB row, **then** it should delete the existing related ArgoCD Secret for Private Repository

Logs:

```
ERROR	row no longer exists in the database	{"operationRow": "test-operation-3", "resource-id": "test-my-repo-creds-3", "error": "cannot find repository credentials: pg: no rows in result set"}
INFO	Deleting Argo CD Secret (leftover) that is missing a DB Entry	{"operationRow": "test-operation-3", "secret": "test-fake-obsolete-secret", "namespace": "gitops-service-argocd"}
INFO	Argo CD Secret (leftover) has been successfully deleted	{"operationRow": "test-operation-3", "secret": "test-fake-obsolete-secret", "namespace": "gitops-service-argocd"}
UPDATE "operation" AS "op" SET "instance_id" = 'test-fake-engine-instance', "resource_id" = 'test-my-repo-creds-3', "operation_owner_user_id" = 'test-repocred-user-id', "resource_type" = 'RepositoryCredentials', "created_on" = '2022-07-11 16:04:53.837382+00:00:00', "last_state_update" = '2022-07-11 16:04:53.867065+00:00:00', "state" = 'Completed', "human_readable_state" = NULL, "seq_id" = 276 WHERE "op"."operation_id" = 'test-operation-3'
1.657555493870097e+09	INFO	Updated Operation state for 'test-operation-3', state is: Completed
```

**Scenario 4**
**When** Operation DB row and its corresponding Operation CR, point to a missing RepositoryCredentials DB row, **then** it should delete all the existing related ArgoCD Secrets for Private Repository.

```
ERROR	row no longer exists in the database	{"operationRow": "test-operation-4", "resource-id": "test-my-repo-creds-4", "error": "cannot find repository credentials: pg: no rows in result set"}
ERROR	SEVERE: unexpected number (more than one) of related ArgoCD secrets	{"operationRow": "test-operation-4", "length": 2}
INFO	Deleting Argo CD Secret (leftover) that is missing a DB Entry	{"operationRow": "test-operation-4", "secret": "test-fake-obsolete-secret-1", "namespace": "gitops-service-argocd"}
INFO	Argo CD Secret (leftover) has been successfully deleted	{"operationRow": "test-operation-4", "secret": "test-fake-obsolete-secret-1", "namespace": "gitops-service-argocd"}
INFO	Deleting Argo CD Secret (leftover) that is missing a DB Entry	{"operationRow": "test-operation-4", "secret": "test-fake-obsolete-secret-2", "namespace": "gitops-service-argocd"}
INFO	Argo CD Secret (leftover) has been successfully deleted	{"operationRow": "test-operation-4", "secret": "test-fake-obsolete-secret-2", "namespace": "gitops-service-argocd"}
UPDATE "operation" AS "op" SET "instance_id" = 'test-fake-engine-instance', "resource_id" = 'test-my-repo-creds-4', "operation_owner_user_id" = 'test-repocred-user-id', "resource_type" = 'RepositoryCredentials', "created_on" = '2022-07-11 16:08:24.094478+00:00:00', "last_state_update" = '2022-07-11 16:08:24.124501+00:00:00', "state" = 'Completed', "human_readable_state" = NULL, "seq_id" = 277 WHERE "op"."operation_id" = 'test-operation-4'
1.657555704127841e+09	INFO	Updated Operation state for 'test-operation-4', state is: Completed
```

**Scenario 5**
**When** Operation DB row and its corresponding Operation CR, point to a missing RepositoryCredentials DB row, **then** it should do nothing (NOP).

logs:

```
ERROR	row no longer exists in the database	{"operationRow": "test-operation-5", "resource-id": "test-my-repo-creds-5", "error": "cannot find repository credentials: pg: no rows in result set"}
UPDATE "operation" AS "op" SET "instance_id" = 'test-fake-engine-instance', "resource_id" = 'test-my-repo-creds-5', "operation_owner_user_id" = 'test-repocred-user-id', "resource_type" = 'RepositoryCredentials', "created_on" = '2022-07-11 16:10:01.188757+00:00:00', "last_state_update" = '2022-07-11 16:10:01.214953+00:00:00', "state" = 'Completed', "human_readable_state" = NULL, "seq_id" = 278 WHERE "op"."operation_id" = 'test-operation-5'
1.6575558012171679e+09	INFO	Updated Operation state for 'test-operation-5', state is: Completed
```

### CODE REVIEW

File: cluster-agent/controllers/managed-gitops/eventloop/repocred.go [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+
+	return noRetry, nil
+}
+
+// processOperation_RepositoryCredentials processes the given operation as a RepositoryCredentials operation.
+// It returns true if the operation should be retried, and false otherwise.
+// It returns an error if there was an error processing the operation.
+func processOperation_RepositoryCredentials(ctx context.Context, dbOperation db.Operation, crOperation operation.Operation, dbQueries db.DatabaseQueries,
+	argoCDNamespace corev1.Namespace, eventClient client.Client, l logr.Logger) (bool, error) {
+	const retry, noRetry = true, false
```
Comment (jgwest) [inline cluster-agent/controllers/managed-gitops/eventloop/repocred.go:93]: :tada:

**Review by jgwest — CHANGES_REQUESTED** [review]
Looks great, thanks Panos! Just some minor tweaks, plus a rebase needed, then G2G!

File: cluster-agent/controllers/managed-gitops/eventloop/repocred_test.go [hunk 2]
```diff
@@ -0,0 +1,825 @@
...
+				Expect(string(secret.Data["username"])).Should(Equal(repositoryCredential.AuthUsername))
+				Expect(string(secret.Data["password"])).Should(Equal(repositoryCredential.AuthPassword))
+				Expect(secret.Labels[controllers.RepoCredDatabaseIDLabel]).Should(Equal(repositoryCredential.RepositoryCredentialsID))
+
+				By(" --- checking Operation DB status ---")
+				err = dbq.GetOperationById(ctx, operationDB)
+				Expect(err).To(BeNil())
+				Expect(operationDB.State).Should(Equal(db.OperationState_Completed))
+			})
+			//It("If there is a valid Operation ID for a problematic ArgoCD secret, fix the secret", func() {
```
[the remark this answers is not in the document]
Comment (drpaneas) [reply cluster-agent/controllers/managed-gitops/eventloop/repocred_test.go:204]: This commented section has to be removed, sorry I forgot about it (Goland IDE tends to auto-wrap big comment section, so I didn't spot this). This "it" it actually the Scenario 2 which is under test already, so all fine.

**Review by jgwest — APPROVED** [review]
Looks great, thanks @drpaneas!

**Pull request merged**